### ISSUE TITLE
Regression in 0.10: fails to parse alphanumeric identifier with leading digit

### ISSUE THREAD

**Issue by dtolnay** [steveklabnik/semver-parser#55]
For example the version of this tower-make release: https://crates.io/crates/tower-make/0.3.0-alpha.2a

#### semver-parser = "0.9"&emsp;(correct behavior)

```rust
fn main() {
    let input = "= 0.3.0-alpha.2a";
    match semver_parser::range::parse(input) {
        Ok(version_req) => println!("{:#?}", version_req),
        Err(error) => eprintln!("{}", error),
    }
}
```

```console
VersionReq {
    predicates: [
        Predicate {
            op: Ex,
            major: 0,
            minor: Some(
                3,
            ),
            patch: Some(
                0,
            ),
            pre: [
                AlphaNumeric(
                    "alpha",
                ),
                AlphaNumeric(
                    "2a",
                ),
            ],
        },
    ],
}
```

#### semver-parser = "0.10"&emsp;(fails to parse)

```rust
fn main() {
    let input = "= 0.3.0-alpha.2a";
    match input.parse::<semver_parser::RangeSet>() {
        Ok(range_set) => println!("{:#?}", range_set),
        Err(error) => eprintln!("{}", error),
    }
}
```

```console
 --> 1:16
  |
1 | = 0.3.0-alpha.2a
  |                ^---
  |
  = expected EOI or logical_or
```

---

As far as I can tell nothing in https://semver.org would make this an invalid semver version. The text doesn't seem to mind leading nonzero digit in alphanumeric identifiers:

> A pre-release version MAY be denoted by appending a hyphen and a series of dot separated identifiers immediately following the patch version. Identifiers MUST comprise only ASCII alphanumerics and hyphens \[0-9A-Za-z-\]. Identifiers MUST NOT be empty. Numeric identifiers MUST NOT include leading zeroes.

and the grammar allows it:

<pre>
&lt;valid semver&gt; ::= &lt;version core&gt; "-" &lt;pre-release&gt; | ...

&lt;pre-release&gt; ::= &lt;dot-separated pre-release identifiers&gt;

&lt;dot-separated pre-release identifiers&gt; ::= &lt;pre-release identifier&gt; | &lt;pre-release identifier&gt; "." &lt;dot-separated pre-release identifiers&gt;

&lt;pre-release identifier&gt;::= &lt;alphanumeric identifier&gt; | &lt;numeric identifier&gt;

<b>&lt;alphanumeric identifier&gt; ::= &lt;non-digit&gt;</b>
<b>                            | &lt;non-digit&gt; &lt;identifier characters&gt;</b>
<b>                            | &lt;identifier characters&gt; &lt;non-digit&gt;</b>
<b>                            | &lt;identifier characters&gt; &lt;non-digit&gt; &lt;identifier characters&gt;</b>

&lt;identifier characters&gt; ::= &lt;identifier character&gt; | &lt;identifier character&gt; &lt;identifier characters&gt;

&lt;identifier character&gt; ::= &lt;digit&gt; | &lt;non-digit&gt;
</pre>

In particular note that `<alphanumeric identifier>` just means there is a non-digit *somewhere*, whether it's the only character, or the first character, or the last character, or some character in the middle. I would expect `2a` to match as an `<alphanumeric identifier>` consisting of `<identifier character>` followed by `<non-digit>`.

**Comment by k0pernicus** [comment]
Hi @dtolnay ,

After investigating on latest version (0.10.2), it seems that your report has been fixed now.

I introduced, in a branch (https://github.com/k0pernicus/semver-parser/tree/introduce_unit_test_for_issue_55), a new unit test to perform the parser on the same type of input: `1.2.3-alpha.2a`.  
The unit test is called `parse_prerelease_composed_alphanumeric`, in `src/version.rs`:

```rust
#[test]
fn parse_prerelease_composed_alphanumeric() {
    let version = "1.2.3-alpha.2a";
    let parsed = version::parse(version).unwrap();
    let expected_pre = vec![
        Identifier::AlphaNumeric(String::from("alpha")),
        Identifier::AlphaNumeric(String::from("2a")),
    ];
    assert_eq!(expected_pre, parsed.pre);
}
```

Based on this unit test, the unit test passed:
> test version::tests::parse_prerelease_composed_alphanumeric ... ok

**Comment by k0pernicus** [comment]
**EDIT:** sorry, I did not fully understood your issue...

After running the debugger on your code, it seems the issue is coming from the `range_set` function in `src/generated.rs` file.  
The code is too complex for me to try to look at it and fix it, as it seems it is generated code...

Sorry again for the previous comment.

**Comment by steveklabnik** [comment]
Yes, `2a` is a valid identifier here, this is a bug.

**Comment by Enet4** [comment]
I made my own attempt at fixing this (so that deps.rs could update its semver dependency, see deps-rs/deps.rs#63), but unfortunately I couldn't tweak the grammar in a way that would fix this while preventing regressions.

The specification for what makes a pre-release identifier (`part`) is somewhat at fault right now.

The current definition ([semver.pest](https://github.com/steveklabnik/semver-parser/blob/a0538ca237169beca9ffdbe5e224b79a9c1197a2/src/semver.pest)) is:
```
part       = { nr | ("-" | '0' .. '9' | 'A' .. 'Z' | 'a' .. 'z')+ }
```

If the parser finds a digit, it assumes that the part is a number and stops once it reaches other non-digit symbol. This makes it fail on alphanumerical identifiers with leading digits.

```
part       = { ("-" | '0' .. '9' | 'A' .. 'Z' | 'a' .. 'z')+ }
```

This is probably what `semver-parser` had before. It accepts and consumes the full identifier, but also accepts numeric identifiers with a leading 0.

I suspect that a more accurate parser would have to peek deep into the characters of the consumed part before assuming that it is a number. Hopefully this wouldn't be too hard to solve for someone experienced in grammar specifications. This or just check each identifier after the fact, simplifying the parser but introducing an additional step.

**Comment by dtolnay** [comment]
This is fixed in semver 1.0.0.

```rust
fn main() {
    println!("{:#?}", semver::VersionReq::parse("= 0.3.0-alpha.2a").unwrap());
}
```
